### ISSUE TITLE
Compute uptime reports

### ISSUE THREAD

**Issue by lars-t-hansen** [NAICNO/Jobanalyzer#104]
The [requirements doc](https://github.com/NAICNO/Jobanalyzer/blob/main/REQUIREMENTS.md) defines uptime as the time when the system is available for computation - it is not "up" if users are locked out for system management reasons.  I don't know how to make that distinction yet, but as an approximation we can define uptime as the time when the system is running sonar data collection.  Sonalyze needs to have an `uptime` command that prints an uptime report based on the sonar data.  The output should probably be compact, something like
```
system-name system-came-up-timestamp system-went-down-timestamp
```
where the last record (if sorted by system-came-up-timestamp) will have a sentinel value for system-went-down-timestamp.

This being sonalyze, the data would be subject to the usual filtering and would have the usual formatting options (text, csv, json).

The requirements currently ask only for historical uptime, which means that the uptime report may be up to an hour old - computed and uploaded at the same time as the load report.

The way we compute uptime is probably by looking at the log, and if there's a gap that is more than the monitoring window then we can assume the system is down.  This means some assumptions about the monitoring window must be passed to sonalyze, most likely.

There's at least one unknown here, which is the precise behavior of Sonar when nvidia-smi times out (which happens often enough to be annoying, see https://github.com/NordicHPC/sonar/issues/92).  We need to make sure a report is still produced in this case.

(I think that if the requirements evolve to want up-to-date status we should implement that some other way than via Sonar.)

Implementation work

- [x] #124 
- [x] https://github.com/NordicHPC/sonar/issues/107
- [x] Implement the `uptime` verb - design report formats and write code
- [x] Add functionality to naicreport to run the uptime verb and merge the data with that from the load verb
- [x] Mitigate the worst UI problems
- [ ] Implement scripts to generate uptime reports and upload them

Blockers (must land first):

- [x] #124 
- [ ] https://github.com/NordicHPC/sonar/issues/107

**Comment by lars-t-hansen** [comment]
OK, that method to compute uptime doesn't actually work 100%, because Sonar will write no records if there is no activity on the system; there will be gaps in the timeline when the system is idle.

We'll want to either update how `sonar ps` works so that it always writes at least one record, or we'll want to add an option to force it to do so.  Then we can choose whether to always write a synthetic record (probably with a special command name "_sonar_is_up_" and pid=job=0) or to write that only when there is nothing else to write.

(The alternative is to not use sonar for this.  But I think sonar is a good fit and I'll explore that further.)

**Comment by lars-t-hansen** [comment]
(Development on branch larstha-implement-uptime in this repo)

**Comment by lars-t-hansen** [comment]
For the dashboard code, the most useful report format will be:
- only "down" records
- json
- all hosts together in a single file with the usual -minutely, -daily, -weekly, -monthly suffixes

**Comment by lars-t-hansen** [comment]
Live data:
```
$ sonalyze uptime --data-path ~/sonar/data -f2w --fmt=all --host=ml1 --only-down --interval=4
device  host            state  start             end               
host    ml1.hpc.uio.no  down   2023-09-29 12:15  2023-09-29 12:42  
host    ml1.hpc.uio.no  down   2023-09-29 12:50  2023-09-29 13:05  
host    ml1.hpc.uio.no  down   2023-09-29 13:10  2023-09-29 13:20  
gpu     ml1.hpc.uio.no  down   2023-10-04 06:35  2023-10-06 10:05  
host    ml1.hpc.uio.no  down   2023-10-06 10:05  2023-10-06 10:30  
```
and (inserted line breaks)
```
$ sonalyze uptime --data-path ~/sonar/data -f2w --fmt=json,device,host,start,end --host=ml1 --only-down --interval=4
[{"device":"host","host":"ml1.hpc.uio.no","start":"2023-09-29 12:15","end":"2023-09-29 12:42"},
 {"device":"host","host":"ml1.hpc.uio.no","start":"2023-09-29 12:50","end":"2023-09-29 13:05"},
 {"device":"host","host":"ml1.hpc.uio.no","start":"2023-09-29 13:10","end":"2023-09-29 13:20"},
 {"device":"gpu","host":"ml1.hpc.uio.no","start":"2023-10-04 06:35","end":"2023-10-06 10:05"},
 {"device":"host","host":"ml1.hpc.uio.no","start":"2023-10-06 10:05","end":"2023-10-06 10:30"}]
```
The use of interval=4 is an artifact of #136.

**Comment by lars-t-hansen** [comment]
The uptime data (really, downtime data) should probably be folded into the load report on the report generation side, doing it in the client is more hassle than necessary.  naicreport can generate all the data and can perform the merging in a sensible way.  Downtime data can be added as data arrays parallel to the existing data.  (Data volume is becoming an issue but should be addressed separately.)

**Comment by lars-t-hansen** [comment]
Folding it into naicreport worked very well, but the resulting plot makes my eyes bleed.  So probably fix #165 along with this one (or add a hack that is sensible and similar in spirit), and then consider upgrading the plotting strategy somehow.
